crates/dbsp/src/trace/layers/column_layer/mod.rs

```diff
@@ -221,7 +238,7 @@ impl<K, R> ColumnLayer<K, R> {
                 let current_diff =
                     unsafe { &mut *shifter.diffs.as_mut_ptr().add(shifter.processed_len) };
 
-                if !retain(current_key, current_diff) {
+                if !retain(current_key, current_diff) || shifter.processed_len < lower_bound {
```

Comment:
The `lower_bound` stuff can probably be broken out into a single `drop_in_place::<*mut [T]>()` call to drop `0..lower_bound`, probably outside of `process_loop()`